internal/app/utils.go

```diff
@@ -333,13 +333,15 @@ func notifySlack(content string, url string, failure bool, executing bool) bool
 	req, err := http.NewRequest("POST", url, bytes.NewBuffer(jsonStr))
 	if err != nil {
 		log.Errorf("Failed to send slack message: %v", err)
+		return false
 	}
 	req.Header.Set("Content-Type", "application/json")
 
 	client := &http.Client{}
 	resp, err := client.Do(req)
 	if err != nil {
 		log.Errorf("Failed to send notification to slack: %v", err)
+		return false
 	}
 	defer resp.Body.Close()
```

Comment:
You should move `defer resp.Body.Close()` before the return (checking resp != nil, of course)
```
  if resp != nil {
    defer resp.Body.Close()
  }
```

Reply:
I checked http.Client.Do source code, and when it returns error, resp is either nil or already closed, so no need to move resp.Body.Close

Reply:
Makes sense, return values are well documented as well https://golang.org/pkg/net/http/#Client.Do